### PR TITLE
health, net/tlsdial: add healthcheck for self-signed cert

### PR DESCRIPTION
**Pull request by andrew-d**
When we make a connection to a server, we previously would verify with the system roots, and then fall back to verifying with our baked-in Let's Encrypt root if the system root cert verification failed.

We now explicitly check for, and log a health error on, self-signed certificates. Additionally, we now always verify against our baked-in Let's Encrypt root certificate and log an error if that isn't successful. We don't consider this a health failure, since if we ever change our server certificate issuer in the future older non-updated versions of Tailscale will no longer be healthy despite being able to connect.

Updates #3198

Change-Id: I00be5ceb8afee544ee795e3c7a2815476abc4abf
Signed-off-by: Andrew Dunham <andrew@du.nham.ca>

### CODE REVIEW

File: health/health.go [hunk 1]
```diff
@@ -476,6 +489,9 @@ func overallErrorLocked() error {
 	if err := envknob.ApplyDiskConfigError(); err != nil {
 		errs = append(errs, err)
 	}
+	for serverName, err := range tlsConnectionErrors {
+		errs = append(errs, fmt.Errorf("TLS connection error for %q: %v", serverName, err))
```
Comment (bradfitz) [inline health/health.go:493]: %w instead of %v?

File: net/tlsdial/tlsdial.go [hunk 2]
```diff
@@ -99,6 +115,12 @@ func Config(host string, base *tls.Config) *tls.Config {
 	return conf
 }
 
+func certIsSelfSigned(cert *x509.Certificate) bool {
```
Comment (bradfitz) [inline net/tlsdial/tlsdial.go:118]: oh, this is tighter than I imagined it'd be at first reading of the call site.

or does this also cover signatures with internal corp CAs?

File: net/tlsdial/tlsdial.go [hunk 3]
```diff
@@ -79,18 +92,21 @@ func Config(host string, base *tls.Config) *tls.Config {
...
+		// so we can log an informational message. This is useful for
+		// detecting SSL MiTM.
 		opts.Roots = bakedInRoots()
-		_, err := cs.PeerCertificates[0].Verify(opts)
+		_, bakedErr := cs.PeerCertificates[0].Verify(opts)
 		if debug() {
-			log.Printf("tlsdial(bake %q): %v", host, err)
+			log.Printf("tlsdial(bake %q): %v", host, bakedErr)
+		} else if bakedErr != nil {
+			log.Printf("tlsdial: server cert for %q is not a Let's Encrypt cert", host)
```
Comment (bradfitz) [inline net/tlsdial/tlsdial.go:104]: this'll be spammy for e.g. all the headscale users in China or anybody running a custom derper with a non-LE cert

Comment (andrew-d) [reply net/tlsdial/tlsdial.go:104]: Good point; let me rate-limit this.

File: net/tlsdial/tlsdial.go (hunk 2 above)
Comment (andrew-d) [reply net/tlsdial/tlsdial.go:118]: It doesn't, for now; as a follow-up I want to import all public root CAs and verify against that set so we can detect non-public CA certs.

File: net/tlsdial/tlsdial.go (hunk 3 above)
Comment (andrew-d) [reply net/tlsdial/tlsdial.go]: We now log only once per `host` per process invocation; PTAL? I could switch this to "every hour/day/etc." instead if you'd rather.

File: net/tlsdial/tlsdial.go [hunk 4]
```diff
@@ -79,18 +97,23 @@ func Config(host string, base *tls.Config) *tls.Config {
...
+		// detecting SSL MiTM.
 		opts.Roots = bakedInRoots()
-		_, err := cs.PeerCertificates[0].Verify(opts)
+		_, bakedErr := cs.PeerCertificates[0].Verify(opts)
 		if debug() {
-			log.Printf("tlsdial(bake %q): %v", host, err)
+			log.Printf("tlsdial(bake %q): %v", host, bakedErr)
+		} else if bakedErr != nil {
+			if _, loaded := tlsdialWarningPrinted.LoadOrStore(host, true); !loaded {
+				log.Printf("tlsdial: server cert for %q is not a Let's Encrypt cert", host)
```
Comment (bradfitz) [inline net/tlsdial/tlsdial.go:110]: get the word "error" or "warn" in this string for grepping ease?

Comment (bradfitz) [reply net/tlsdial/tlsdial.go:110]: I guess it's an error if errSys is non-nil and warn otherwise?

File: net/tlsdial/tlsdial.go [hunk 5]
```diff
@@ -33,6 +36,11 @@ var sslKeyLogFile = os.Getenv("SSLKEYLOGFILE")
 
 var debug = envknob.RegisterBool("TS_DEBUG_TLS_DIAL")
 
+// tlsdialWarningPrinted tracks whether we've printed a warning about a given
+// hostname already, to avoid log spam for users with custom DERP servers,
+// Headscale, etc.
+var tlsdialWarningPrinted sync.Map
```
Comment (bradfitz) [inline net/tlsdial/tlsdial.go:42]: for untyped collections, always document the keys + value types in a little same-line comment

**Review by bradfitz — APPROVED** [review]

File: net/tlsdial/tlsdial.go (hunk 4 above)
Comment (andrew-d) [reply net/tlsdial/tlsdial.go]: Added the word `warning` or `error` depending on that case.

**Pull request merged**